[#2978] GeneralAdmin: Disallow "modify" of "user" of type "rodsgroup" (main)

alanking opened the pull request:
In service of #2978

Companion PR: https://github.com/irods/irods_client_icommands/pull/532

Unit tests passed. Core tests still running. Will notify once those pass.

alanking commented:
Remaining clang-tidy issues are #8155-like.

korydraughn reviewed (commented):
Looking good so far.

alanking commented on server/api/src/rsGeneralAdmin.cpp:
```diff
@@ -763,12 +764,29 @@ _rsGeneralAdmin( rsComm_t *rsComm, generalAdminInp_t *generalAdminInp ) {
 
     if ( strcmp( generalAdminInp->arg0, "add" ) == 0 ) {
         if ( strcmp( generalAdminInp->arg1, "user" ) == 0 ) {
+            // If the client requested to "add" a "user" of type "rodsgroup", return an error because one can now "add"
+            // a "group".
+            if (nullptr != generalAdminInp->arg3 && 0 == std::strcmp(generalAdminInp->arg3, "rodsgroup")) {
+                constexpr auto ec = SYS_NOT_ALLOWED;
+                constexpr const char* msg =
+                    "Using 'add' with 'user' and type 'rodsgroup' is not allowed. Use 'add' with 'group' instead.";
+                addRErrorMsg(&rsComm->rError, ec, msg);
+                log_api::info("{}: {}", __func__, msg);
+                return ec;
+            }
```
`iadmin mkgroup` still exists in iRODS 4 clients and is the way to create groups in an iRODS 5 zone. This change only prevents `iadmin mkuser` from creating groups, but iRODS 4 clients can continue to create groups in iRODS 5 zones via `iadmin mkgroup`.

korydraughn replied:
Ah, I missed that.

Thanks.

alanking commented:
Core tests passed. I'm going to run `test_iadmin` again after addressing review comments since nothing is really changing in the C++.

alanking commented on lib/administration/user/src/user_administration.cpp:
```diff
@@ -121,9 +121,24 @@ namespace irods::experimental::administration::NAMESPACE_IMPL
+        input.arg1 = "group";
+#else
+        // If this is not the server side, what we provide to the API depends on the version of the server with which
+        // the client is connected. For 4.3.4 and later, the API can add a "group" and do not need to provide a user
```
Updated

alanking commented on scripts/irods/test/test_iadmin.py:
```diff
@@ -2437,12 +2437,30 @@ def test_moduser_type_rodsadmin_rodsuser(self):
-        self.assertEqual('rodsuser', lib.get_user_type(self.admin, self.username))
+        self.assertEqual(user_type, lib.get_user_type(self.admin, target_username))
+        self.admin.assert_icommand(
+            ["iadmin", "moduser", target_username, 'type', "rodsgroup"], 'STDERR', 'SYS_NOT_ALLOWED')
```
Done

alanking commented on scripts/irods/test/test_iadmin.py:
```diff
@@ -2437,12 +2437,30 @@ def test_moduser_type_rodsadmin_rodsuser(self):
+        """Test modifying the rodsadmin's type to a group (not allowed)."""
+        self.moduser_to_rodsgroup_is_not_allowed_test_impl(self.admin.username, "rodsadmin")
+
+    def test_moduser_type_groupadmin_rodsgroup(self):
```
Done

alanking commented on scripts/irods/test/test_iadmin.py:
```diff
@@ -2616,142 +2633,87 @@ def tearDownClass(self):
-        self.admin.assert_icommand(['iadmin', 'moduser', self.group, 'type', 'rodsadmin'],
-                                   'STDERR', 'SYS_NOT_ALLOWED')
-        self.assertEqual('rodsgroup', lib.get_user_type(self.admin, self.group))
+    def test_moduser_type_invalid_user_type_is_not_allowed(self):
```
Done

alanking commented on scripts/irods/test/test_iadmin.py:
```diff
@@ -2616,142 +2633,87 @@ def tearDownClass(self):
-
+    def test_mkuser_is_not_allowed_with_groups__issue_2978(self):
+        """Test that mkuser with type rodsgroup is not allowed."""
+        self.admin.assert_icommand(["iadmin", "mkuser", self.group, "rodsgroup"], "STDERR", "SYS_NOT_ALLOWED")
```
Done

alanking commented on server/api/src/rsGeneralAdmin.cpp:
```diff
@@ -763,12 +764,29 @@ _rsGeneralAdmin( rsComm_t *rsComm, generalAdminInp_t *generalAdminInp ) {
+                constexpr const char* msg =
+                    "Using 'add' with 'user' and type 'rodsgroup' is not allowed. Use 'add' with 'group' instead.";
+                addRErrorMsg(&rsComm->rError, ec, msg);
+                log_api::info("{}: {}", __func__, msg);
```
Done

alanking commented on server/api/src/rsGeneralAdmin.cpp:
```diff
@@ -763,12 +764,29 @@ _rsGeneralAdmin( rsComm_t *rsComm, generalAdminInp_t *generalAdminInp ) {
+            if (nullptr != generalAdminInp->arg3 && 0 == std::strcmp(generalAdminInp->arg3, "rodsgroup")) {
+                constexpr auto ec = SYS_NOT_ALLOWED;
+                constexpr const char* msg =
+                    "Using 'add' with 'user' and type 'rodsgroup' is not allowed. Use 'add' with 'group' instead.";
```
Done

alanking commented:
`test_iadmin` passed. I added one more assertion to an existing test based on this statement from #2978:

> If you moduser a group to give it a password, you can log in as that group. Also, on creation, groups are added as members of themselves. Neither of these behaviors are intentional.

We already have tests around setting passwords on groups, so I added an assertion in a test which creates a group to ensure that the group is not a member of itself.

I think this is ready, barring any additional review comments, of course.

korydraughn reviewed (commented):
Looks good to me.

Squash to taste.

alanking commented:
Squashed

korydraughn reviewed (approved):
Pound it.
